ledger-service/http-json/src/main/scala/com/digitalasset/http/json/JsonProtocol.scala

```diff
@@ -451,8 +460,58 @@ object JsonProtocol extends JsonProtocolLow {
   implicit def OkResponseFormat[R: JsonFormat]: RootJsonFormat[domain.OkResponse[R]] =
     jsonFormat3(domain.OkResponse[R])
 
+  implicit val ResourceInfoDetailFormat: RootJsonFormat[ResourceInfoDetail] = jsonFormat2(
+    ResourceInfoDetail
+  )
+  implicit val ErrorInfoDetailFormat: RootJsonFormat[ErrorInfoDetail] = jsonFormat2(
+    ErrorInfoDetail
+  )
+
+  implicit val RetryInfoDetailFormat: RootJsonFormat[RetryInfoDetail] =
+    new RootJsonFormat[RetryInfoDetail] {
+      override def write(obj: RetryInfoDetail): JsValue = JsObject(
+        "duration" -> JsNumber(obj.duration.toNanos)
+      )
+      override def read(json: JsValue): RetryInfoDetail = json match {
+        case JsObject(fields) =>
+          fields
+            .get("duration")
+            .collect { case JsNumber(nanos) =>
+              val duration = scala.concurrent.duration.Duration.fromNanos(nanos.toLongExact)
+              RetryInfoDetail(duration)
+            }
+            .getOrElse(deserializationError("Expected field duration of type number"))
+        case _ => deserializationError("Expected an object with field duration of type number")
+      }
+    }
+
+  implicit val RequestInfoDetailFormat: RootJsonFormat[RequestInfoDetail] = jsonFormat1(
+    RequestInfoDetail
+  )
+
+  implicit val ErrorDetailsFormat: RootJsonFormat[ErrorDetail] =
+    new RootJsonFormat[ErrorDetail] {
+      override def write(obj: ErrorDetail): JsValue = obj match {
+        case a: ResourceInfoDetail => ResourceInfoDetailFormat.write(a)
+        case a: ErrorInfoDetail => ErrorInfoDetailFormat.write(a)
+        case a: RetryInfoDetail => RetryInfoDetailFormat.write(a)
+        case a: RequestInfoDetail => RequestInfoDetailFormat.write(a)
+      }
+      override def read(json: JsValue): ErrorDetail =
+        Try(ResourceInfoDetailFormat.read(json))
```

Comment:
This seems error prone, can we add a discriminator column in the error detail json, so we can just key on that?